### PR TITLE
added STAR aligner package

### PR DESCRIPTION
**Pull request by JusticeForMikeBrown**
OK this compiles but does not install properly.

The `make install` phase simply moves the compiled binaries to `../bin `directory under the staging directory.  I cannot discern from the developer's [docs](https://github.com/alexdobin/STAR/blob/master/README.md) how I can pass Spack prefix to `make install`.

To get around that I added:

```
shutil.move("../bin/STAR", prefix)
shutil.move("../bin/STARlong",  prefix)

```

Which puts the binaries into the prefix

```
user@node103[~/repos/spack]$ spack find -p star
==> 1 installed packages.
-- linux-rhel6-x86_64 / gcc@6.3.0 -------------------------------
    star@2.5.3a  /pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/star-2.5.3a-pptd3lygyhgy3p5qgmzkmkgw4taq7mkq
user@node103[~/repos/spack]$ ls -ltr /pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/star-2.5.3a-pptd3lygyhgy3p5qgmzkmkgw4taq7mkq
total 3.7M
-rwxr-xr-x 1 user physbio 1.9M May 26 15:00 STARlong
-rwxr-xr-x 1 user physbio 1.9M May 26 15:00 STAR
```
Though `spack load` does not add that to `$PATH`.  I read the docs on `Package()` class though I must have missed a step.  Can you point me in the right direction as I am hoping it's some trivial thing.

### CODE REVIEW

**Comment by JusticeForMikeBrown** [comment]
The problem seems to be that the module file lacks prepend-path for $PATH but only contains 

> prepend-path CMAKE_PREFIX_PATH

File: var/spack/repos/builtin/packages/star/package.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    homepage = "https://github.com/alexdobin/STAR"
+    url      = "https://github.com/alexdobin/STAR/archive/2.5.3a.tar.gz"
+
+    version('2.5.3a', 'baf8d1b62a50482cfa13acb7652dc391')
+
+    def install(self, spec, prefix):
+        source_directory = join_path(self.stage.source_path, 'source')
+
+        with working_dir(source_directory):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/star/package.py]: You could just do `working_dir('source')` if you want, it doesn't have to be an absolute path.

File: var/spack/repos/builtin/packages/star/package.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            make('STAR', 'STARlong')
+            make('install', 'STAR', 'STARlong')
+            shutil.move("../bin/STAR", prefix)
+            shutil.move("../bin/STARlong", prefix)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/star/package.py]: As you've discovered, the `Makefile` contains:
```make
.PHONY: install
install:
        'mv' STAR STARlong ../bin
```
This is pretty much useless. Don't bother calling `make install`, just copy the files using the `install()` command.

**Comment by JusticeForMikeBrown** [comment]
Yay

```
user@node103[~/repos/spack]$ spack load star
user@node103[~/repos/spack]$ which STAR
/pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/star-2.5.3a-pptd3lygyhgy3p5qgmzkmkgw4taq7mkq/STAR
user@node103[~/repos/spack]$ STAR | head
Usage: STAR  [options]... --genomeDir REFERENCE   --readFilesIn R1.fq R2.fq
Spliced Transcripts Alignment to a Reference (c) Alexander Dobin, 2009-2015

### versions
versionSTAR             020201
    int>0: STAR release numeric ID. Please do not change this value!
versionGenome           020101 020200
    int>0: oldest value of the Genome version compatible with this STAR release. Please do not change this value!

### Parameter Files
```

File: var/spack/repos/builtin/packages/star/package.py (hunk 2 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/star/package.py]: ZING!

**Comment by adamjstewart** [comment]
Install the files to `prefix.bin`, that's where binaries usually go.

**Comment by JusticeForMikeBrown** [comment]
@adamjstewart will do, thanks!

**Comment by JusticeForMikeBrown** [comment]
I installed with

`%gcc@6.3.0 cxxflags="-flto -march=native" ldflags="-flto"`

Can I make user get prompted if that version of GCC isn't being used?

Should I add those compiler flags or leave that up to the person who's installing?

**Comment by adamjstewart** [comment]
> Can I make user get prompted if that version of GCC isn't being used?

Only if you use Lmod.

> Should I add those compiler flags or leave that up to the person who's installing?

Depends on how generic they are and whether or not they are required to build. I wouldn't add `-march=native`. It won't work if you aren't building with GCC, and it won't work if you are cross-compiling. I don't recommend adding architecture-specific flags.

File: var/spack/repos/builtin/packages/star/package.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+class Star(Package):
+    """STAR is an ultrafast universal RNA-seq aligner."""
+
+    homepage = "https://github.com/alexdobin/STAR"
+    url      = "https://github.com/alexdobin/STAR/archive/2.5.3a.tar.gz"
+
+    version('2.5.3a', 'baf8d1b62a50482cfa13acb7652dc391')
+
+    def setup_environment(self, spack_env, run_env):
+        run_env.prepend_path('PATH', join_path(self.prefix.bin))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/star/package.py]: You don't need this. `prefix.bin` is automatically added to PATH.

File: var/spack/repos/builtin/packages/star/package.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    def install(self, spec, prefix):
+        join_path(self.stage.source_path, 'source')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/star/package.py]: You don't need this line anymore.

**Comment by hartzell** [comment]
:+1: -- This was one of the next ones on my list.  **Thanks!**

File: var/spack/repos/builtin/packages/star/package.py (hunk 3 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/star/package.py]: @adamjstewart I didn't see it in module file until I added this line.  Let me comment out and verify again.

File: var/spack/repos/builtin/packages/star/package.py (hunk 4 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/star/package.py]: You rite

File: var/spack/repos/builtin/packages/star/package.py (hunk 3 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/star/package.py]: @adamjstewart You rite per usual

**Comment by JusticeForMikeBrown** [comment]
@hartzell No problem!

I will now create another PR for STAR-Fusion.  I wanted to include it in this package but it's perl based with different dependencies.

**Comment by adamjstewart** [comment]
> I wanted to include it in this package but it's perl based with different dependencies.

The general rule is "one package per tarball", so if it comes in the same tarball it should be the same package, and if not it should be a separate package.

**Comment by JusticeForMikeBrown** [comment]
It's a separate tarball so I will do another PR.

**Pull request merged**